### ISSUE TITLE
invalid Wi-Fi popup on macOS

### ISSUE THREAD

**Issue by thelazyoxymoron** [koreader/koreader#12203]
* KOReader version: `2024.07-25 (1.0)` (taken from this commit artifact: [5affb68](https://github.com/koreader/koreader/commit/5affb68233fda994150df1da3d0b573402f20842))
* Device: MacOS (M1)

#### Issue
* Go to settings -> Network
* There's no setting for WiFi connection there
* Go to OPDS Catalog -> Click on any of the default entries (e.g. Project Gutenberg) -> Get "Do you want to turn on Wi-Fi" popup -> Click on "Turn on" -> Nothing happens

**Comment by benoit-pierre** [comment]
If you start koreader from a terminal, after setting the environment variable `KO_MULTIUSER` to non-empty string (e.g. `export KO_MULTIUSER=1`), is the behavior different?

**Comment by thelazyoxymoron** [comment]
Yes. I get the popup and when pressed "Turn on", nothing happens. If I click on any of the OPDS libraries again, I get the following warning in terminal:

```
07/22/24-12:07:46 WARN  NetworkMgr:promptWifiOn: A previous connection attempt is still ongoing!
```

**Comment by benoit-pierre** [comment]
Never mind, that should already be the case anyway:
- https://github.com/koreader/koreader-base/blob/master/osx_loader.c#L25
- https://github.com/koreader/koreader-base/blob/master/osx_loader.c#L60

**Comment by benoit-pierre** [comment]
Also, please provide the full koreader version (including the commit hash).

**Comment by thelazyoxymoron** [comment]
Using the artifact built by this commit hash: [5affb68](https://github.com/koreader/koreader/commit/5affb68233fda994150df1da3d0b573402f20842). Updated in my original post as well.

**Comment by benoit-pierre** [comment]
I image there's no `/proc/net/route` on macOS?

Cf.: https://github.com/koreader/koreader/blob/master/frontend/device/generic/device.lua#L776

**Comment by Frenzie** [comment]
I'm not sure if /proc exists on Mac at all; I might check later.

In the meantime, this might be of some interest https://stackoverflow.com/a/11265543

**Comment by thelazyoxymoron** [comment]
I'm afraid not. I was able to get default gateway via following two methods:

* Using `/sbin/route -n get default`:
```
   route to: default
destination: default
       mask: default
    gateway: 10.1.0.1
  interface: en0
      flags: <UP,GATEWAY,DONE,STATIC,PRCLONING,GLOBAL>
 recvpipe  sendpipe  ssthresh  rtt,msec    rttvar  hopcount      mtu     expire
       0         0         0         0         0         0      1500         0 
```

* Using `/usr/sbin/scutil --nwi`:
```
Network information

IPv4 network interface information
     en0 : flags      : 0x5 (IPv4,DNS)
           address    : 10.1.2.101
           reach      : 0x00000002 (Reachable)

   REACH : flags 0x00000002 (Reachable)

IPv6 network interface information
   No IPv6 states found


   REACH : flags 0x00000000 (Not Reachable)

Network interfaces: en0
```

**Comment by Frenzie** [comment]
PS We don't need to replicate that which we do on the likes of Kobo; if there's an easy way to outsource it that's perfectly fine.

**Comment by benoit-pierre** [comment]
I don't understand why we even try to check if we're connected when the device does not support toggling Wi-Fi:
```diff
 frontend/device/sdl/device.lua | 3 +++
 1 file changed, 3 insertions(+)

diff --git i/frontend/device/sdl/device.lua w/frontend/device/sdl/device.lua
index 1f2d46efe3..eb41f5c792 100644
--- i/frontend/device/sdl/device.lua
+++ w/frontend/device/sdl/device.lua
@@ -396,6 +396,9 @@ end
 function Device:initNetworkManager(NetworkMgr)
     function NetworkMgr:isWifiOn() return true end
     function NetworkMgr:isConnected()
+        if not Device:hasWifiToggle() then
+            return true
+        end
         -- Pull the default gateway first, so we don't even try to ping anything if there isn't one...
         local default_gw = Device:getDefaultRoute()
         if not default_gw then
```

**Comment by benoit-pierre** [comment]
IMHO, we're better off just attempting whatever network access we need (with a sufficiently verbose error message in case of failure).

**Comment by benoit-pierre** [comment]
PS: I get the same behavior when running the emulator with `KO_MULTIUSER=1 ./kodev run -n` and no network access.

**Comment by Frenzie** [comment]
> IMHO, we're better off just attempting whatever network access we need (with a sufficiently verbose error message in case of failure).

It's not, because then you need to wait an eternity for the time out. That's just as true on desktop as anywhere else, lack of polish and a dialog that shouldn't be shown there notwithstanding.

**Comment by benoit-pierre** [comment]
If the user asked to connect, then what's the issue? Unless there's no way to interrupt the wait, then I would argue that's the real problem.

**Comment by Frenzie** [comment]
> Unless there's no way to interrupt the wait, then I would argue that's the real problem.

Yup! ;-) That's why there's Turbo support in the source (although probably no longer quite functional) but that apparently caused some hard to trace crashes, especially on Android. Of course that was quite a few years ago by now.

But either way, it's a lot more helpful to see this:
![image](https://github.com/user-attachments/assets/e301cf18-4d90-4c84-965b-064338be47ff)

Than this:
![Screenshot_20240722_094758](https://github.com/user-attachments/assets/7d6cf3d2-d26b-4561-aa31-2ab91c4b5b49)

**Comment by NiLuJe** [comment]
That doesn't really help when we return false in `isConnected`, though, because NetworkMgr really expects to be able to do something about it, and that's not really the case on !hasWifiToggle platforms :/.

**Comment by NiLuJe** [comment]
TL;DR: I don't really see a better way to handle this than https://github.com/koreader/koreader/issues/12203#issuecomment-2242261416

c.f. https://github.com/koreader/koreader/blob/5affb68233fda994150df1da3d0b573402f20842/frontend/ui/network/manager.lua#L165-L177

I clearly forgot than SDL implemented this, so its whole `Device:initNetworkManager` block can probably go bye-bye and everything will be as well as it can be.
